Making fingerprint part of SDP created by createOffer/Answer

stefhak opened the pull request:
First cut. The key and certificate is hard coded. Tested to work with FF (sending the SDPs from createO/A).

stefhak commented:
Tested to work with FF.

stefhak commented:
I updated a bit. Could someone please review this one (I think there are plenty of things in _crypto_utils.c that are not up to par)?

stefanalund commented:
@pererikb could you have a look at this?

stefhak commented:
@stefanalund, @Rugvip  has already promised to review.

Rugvip commented on transport/owr_transport_agent.c:
```diff
@@ -1158,14 +1158,18 @@ static GstElement *add_dtls_srtp_bin(OwrTransportAgent *transport_agent, guint s
     if (!is_encoder) {
         g_object_get(session, "dtls-certificate", &cert, NULL);
         g_object_get(session, "dtls-key", &key, NULL);
+        g_print("transport_agent cert = %s\nkey = %s\n", cert, key);
```
These should be replaced by GST_DEBUG calls, e.g.

```
GST_CAT_INFO_OBJECT(_owrsession_debug, session, "transport_agent cert = %s\nkey = %s\n", cert, key);
```

Rugvip commented on transport/owr_session.c:
```diff
@@ -148,6 +148,7 @@ static void owr_session_set_property(GObject *object, guint property_id, const G
             g_free(priv->dtls_certificate);
             priv->dtls_certificate = NULL;
         }
+        g_print("cert in owr_session.c: %s\n", priv->dtls_certificate);
```
Should be replaced by GST_DEBUG calls, i.e.

```
GST_DEBUG_OBJECT(session, "cert in owr_session.c: %s\n", priv->dtls_certificate);
```

Rugvip commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,230 @@
+
+  if (errorDetected) {
+    g_print("error detected \n");
+    g_value_set_string(&params[0], "Failure");
```
Perhaps errors should be handled by a 4th argument, and all of these should be set to null. That would open up for giving some more descriptive reasons in the future. We haven't really standardised to a specific error forwarding scheme though, so perhaps we should just keep this as is and maybe change it in the future.

Rugvip commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,230 @@
+
+    g_value_init(&params[0], G_TYPE_STRING);
+    g_value_init(&params[1], G_TYPE_STRING);
+    g_value_set_string(&params[1], pem_cert);
```
This and the _set_string below seem unnecessary.

pearcetm commented:
This PR will be very useful for the project I'm developing - now that it has been reviewed, what kind of timeline are we looking at for getting these changes updated and eventually merged? A few days, a week or two, months... I know nothing is for sure but a rough idea would be really helpful in planning for further development.

stefhak commented:
I'll try to address the comments tomorrow, I think (fingers crossed) we should get this one in within a few days.

stefhak commented:
@Rugvip I addressed your comments (they were good, thanks) except for the error reporting one which I defered to a later error handing update. PTAL.

Rugvip commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,228 @@
+    g_value_init(&params[2], G_TYPE_STRING);
+
+  if (errorDetected) {
+    g_print("error detected \n");
```
In general we should avoid g_print in committed code. All of these g_print statements should be replaced by GStreamer debug logging functions, so that we can show and hide them as needed, and provide some context. Especially this print statement will be very confusing unless you know exactly where you should be looking.

stefhak replied:
I can easily change them all, but need some help. Should I use GST_DEBUG_OBJECT? What object should I tie it do?

stefhak commented:
@Rugvip: have now changed the g_print's to GST_DEBUG's. PTAL.

Rugvip commented:
Just a minor thing, there's no need to mention the file in GST_DEBUG, as it's already a part of the log message, and there's also no need to add a newline. Otherwise, :+1:

stefhak commented:
@Rugvip fixed those two items.

superdump commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,228 @@
+  X509_sign (cert, key_pair, EVP_sha256 ());
+
+  if (!X509_digest(cert, fprint_type, fprint, &fprint_size)) {
+    GST_DEBUG("Error, could not create certificate fingerprint");
```
GST_ERROR()

superdump commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,228 @@
+  char_fprint = g_string_free(string_fprint, FALSE);
+
+  if (!PEM_write_bio_X509 (bio_cert, (X509 *) cert)) {
+    GST_DEBUG("Error, could not write certificate bio");
```
GST_ERROR()

superdump commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,228 @@
+
+
+  if (!PEM_write_bio_PrivateKey (bio_key, (EVP_PKEY *) key_pair, NULL, NULL, 0, 0, NULL)) {
+    GST_DEBUG("Error, could not write PrivateKey bio");
```
GST_ERROR()

superdump commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,228 @@
+
+  len_cert = BIO_read (bio_cert, buffer_cert, GST_DTLS_BIO_BUFFER_SIZE);
+  if (!len_cert) {
+    GST_DEBUG("Error, no cert length");
```
GST_ERROR()

superdump commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,228 @@
+
+  len_key = BIO_read (bio_key, buffer_key, GST_DTLS_BIO_BUFFER_SIZE);
+  if (!len_key) {
+    GST_DEBUG("Error, no key length");
```
GST_ERROR()

superdump commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,228 @@
+    g_value_init(&params[2], G_TYPE_STRING);
+
+  if (errorDetected) {
+    GST_DEBUG("Returning with error");
```
GST_ERROR()

superdump commented on transport/owr_session.c:
```diff
@@ -148,6 +148,7 @@ static void owr_session_set_property(GObject *object, guint property_id, const G
             g_free(priv->dtls_certificate);
             priv->dtls_certificate = NULL;
         }
+        GST_DEBUG_OBJECT(OWR_SESSION(object), "cert in owr_session.c: %s\n", priv->dtls_certificate);
```
As mentioned you don't need to mention the source file in the log message, it will be included.

superdump commented on transport/owr_transport_agent.c:
```diff
@@ -1158,11 +1158,12 @@ static GstElement *add_dtls_srtp_bin(OwrTransportAgent *transport_agent, guint s
             g_object_get(dtls_srtp_bin, "pem", &cert, NULL);
             g_object_set(session, "dtls-certificate", cert, NULL);
             g_object_set(session, "dtls-key", NULL, NULL);
+            g_print("used auto generated cert stuff.. THIS SHOULD NOT HAPPEN!!!!\n");
```
No `g_print`. You can use GST_ERROR() followed by `g_assert_not_reached();`

superdump commented on transport/owr_transport_agent.c:
```diff
@@ -1158,11 +1158,12 @@ static GstElement *add_dtls_srtp_bin(OwrTransportAgent *transport_agent, guint s
         g_object_get(session, "dtls-certificate", &cert, NULL);
         g_object_get(session, "dtls-key", &key, NULL);
-
+        GST_CAT_INFO_OBJECT(_owrsession_debug, session, "transport_agent cert = %s\nkey = %s\n", cert, key);
```
Why does the cert need to be printed in multiple places during normal usage when the code is actually working?

stefhak replied:
I don't think that is necessary, I'll remove this line. Is there anywhere else you see an actual (non debug or error) printout?

superdump commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,228 @@
+#include <openssl/ssl.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+
```
Please add a debug category for owrcrypto, else GST_DEBUG() and such will use the 'default' (literally) category.

Add this here:
```
GST_DEBUG_CATEGORY_EXTERN(_owrcrypto_debug);
#define GST_CAT_DEFAULT _owrcrypto_debug
```

The following at https://github.com/EricssonResearch/openwebrtc/blob/master/owr/owr.c#L72 :
```
GST_DEBUG_CATEGORY(_owrcrypto_debug);
```

And the following here https://github.com/EricssonResearch/openwebrtc/blob/master/owr/owr.c#L225 :
```
    GST_DEBUG_CATEGORY_INIT(_owrcrypto_debug, "owrcrypto", 0,
        "OpenWebRTC Cryptography");
```

superdump commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,228 @@
+
+void owr_crypto_create_crypto_data(OwrCryptoDataCallback callback)
+{
+
```
The style of this file is not in line with the rest of the project. But I can fix that quickly post-merge.

stefhak commented:
Thanks @superdump, very valuable feedback, I'll fix those issues ASAP.

stefhak commented:
@superdump @Rugvip @pererikb pushed an update that hopefully addresses your concerns. PTAL.

superdump commented:
If @Rugvip is happy with the openssl part, I think I'm happy with this. :smile:

stefhak commented:
@Rugvip you think this looks awesome, right? ;)

Rugvip commented:
Yep, as long as it works :)

stefhak commented:
I works fine for me at least. As was noted in a separate thread, bridgeserver.js is locked up while crypto data is generated. That is my bad, and I intend to fix it, but it is for a very short time. I'd propose we merge this, and I generate a separate PR to fix that (but then I'm biased :) ).

stefhak commented:
Reassigning to @superdump - anything more I should do before merging?

superdump commented:
I'll merge it when I have time to fix the style. In the mean time you're welcome to fix the asynchronousness. :smile:

stefhak commented:
@superdump I'm looking into the asynchronousness right now :) I'm reading about stuff like g_idle_add() (which does not seem like the way to go). Any pointers would be welcome!

superdump commented:
`void *`! :laughing:

More seriously, I think you may want something like one of these internal schedule functions: https://github.com/EricssonResearch/openwebrtc/blob/master/owr/owr_private.h#L47

They are somewhat like g_idle_add in that they use an idle GSource with default priority but they make sure to attach to the correct main context. They are implemented here if you want the details: https://github.com/EricssonResearch/openwebrtc/blob/master/owr/owr.c#L414

stefhak commented:
@superdump @pererikb now it's asynchronous

adam-be commented on transport/owr_crypto_utils.c:
```diff
@@ -0,0 +1,274 @@
+  GThread *crypto_worker;
+
+
+  crypto_worker = g_thread_new("crypto_worker", _create_crypto_worker_run, (gpointer)callback);
```
Should we pass along the GThread here and unref it once the work is done?

stefhak commented:
@adam-be adding unreferencing of GThread with latest commit.
@superdump merge?

adam-be commented on bridge/worker/bridgeserver.js:
```diff
@@ -119,6 +119,26 @@ server.onaccept = function (event) {
+                            "certificate": certificate,
+                            "privatekey": privatekey,
+                            "fingerprint": fingerprint
+                    });
```
Code style: Move the { on 132 up to 131 and save one level of indentation

stefanalund commented:
Stay in there Stefan! 😀

stefhak commented:
Fixed Adam's style comment, and removed a couple of white space changes.

stefhak commented:
@superdump did some more styling things. Can you PTAL?

stefhak commented:
@superdump I run "clang-format -style=WebKit" on owr_crypto_utils.c

stefanalund commented:
I really think we should try to land this now @adam-be @pererikb @superdump

superdump commented:
I was working on merging this but it seems that it does not rebase cleanly. There are conflicts with the code in master. Sorry for the inconvenience, but can you rebase on top of master? Then I will merge it.

stefhak commented:
Thanks for working on this. I will rebase.

stefhak commented:
So, I think (fingers crossed) I managed to rebase. FWIW I have re-run the tests, and things still seem to work.

superdump commented:
It is done. :)

stefanalund commented:
Wohooo, this is huge! :smile:

stefhak commented:
:-)

stefanalund commented:
I'm currently working on a major new version of Bowser as well that makes it much more reliable (using WKWebView). Combined with this PR we should soon have a Bowser that is significantly better.

ford-prefect commented:
The assert added in add_dtls_srtp_bin() seems to break test-send-receive.

superdump commented:
Indeed. We'll have to fix that.
